src/chips/sam4l/flashcalw.rs

```diff
@@ -0,0 +1,1103 @@
+            page_buffer:            TakeCell::new([0; FLASH_PAGE_SIZE as usize])
+        }
+    }
+
+    // Helper to read a flashcalw register 
+    fn read_register(&self, key : RegKey) -> u32 {
+        let registers : &mut Registers = unsafe { 
+            mem::transmute(self.registers)
+        };
+        
+        match key {
+            RegKey::CONTROL => {
+                registers.control.get()
+            },
+            RegKey::COMMAND => {
+                registers.command.get()
+            },
+            RegKey::STATUS => {
+                registers.status.get()
+            },
+            RegKey::PARAMETER => {
+                registers.parameter.get()
+            },
+            RegKey::VERSION => {
+                registers.version.get()
+            },
+            RegKey::GPFRHI => {
+                registers.general_purpose_fuse_register_hi.get()
+            },
+            RegKey::GPFRLO => {
+                registers.general_purpose_fuse_register_lo.get()
+            } 
+        }
+    }
+
+
+    pub fn handle_interrupt(&self) {
+        unsafe { 
+            //  mark the controller as ready and clear pending interrupt 
+            self.mark_ready();  
+            nvic::clear_pending(nvic::NvicIdx::HFLASHC);
+        }
+       
+        self.error_status.put(Some(self.get_error_status()));
+        let error_status = self.error_status.take().unwrap();
+        self.error_status.put(Some(error_status));
+
+        //  Since the only interrupt request on is FRDY, a command should have
+        //  either completed or failed at this point.
+
+        // Check for errors and report to Client if there are
+        if error_status != 0 {
+            //reset commands / ready
+            self.current_command.set(Command::None);
+            self.current_state.set(FlashState::Ready);
+            
+            //call command complete with error
+            match error_status {
+                4 =>  {
+                    self.client.map(|value| {
+                        value.command_complete(Error::LockE);
+                    });
+                },
+                8 =>  {
+                    self.client.map(|value| {
+                        value.command_complete(Error::ProgE);
+                    });
+                },
+                12 => {
+                    self.client.map(|value| {
+                        value.command_complete(Error::LockProgE);
+                    });
+                },
+                _ => {}
+            }
+            return;
+        }
+        
+        //  Part of a command succeeded -- continue onto next steps.
+         
+        match self.current_command.get() {
+            Command::Write => {
+                match self.current_state.get() {
+                    FlashState::Unlocking => {
+                        self.current_state.set(FlashState::Erasing);
+                        self.flashcalw_erase_page(self.page.get(), true);
```

Comment:
seems to me like, in general, you're using state variables like the page number in conjunction with particular in progress commands. It's equivalent, and IMO more semantically meaningful, to store these alongside either the `current_command` or `current_state`, e.g.:

```rust
enum Command {
    Write { page: u32 },
    Read { page: u32 },
    ...
}
```